snuba/util.py

```diff
@@ -193,19 +194,41 @@ def wrapper(*args, **kwargs):
 
 
 class Part(NamedTuple):
+    name: str
     date: datetime
     retention_days: int
 
 
-def decode_part_str(part_str: str) -> Part:
-    match = PART_RE.match(part_str)
+class PartSegment(Enum):
+    RETENTION_DAYS = "retention_days"
+    DATE = "date"
+
+
+def decode_part_str(part_str: str, part_format: Sequence[PartSegment]) -> Part:
+    PARTSEGMENT_RE = {
+        PartSegment.DATE: "('(?P<date>\d{4}-\d{2}-\d{2})')",
+        PartSegment.RETENTION_DAYS: "(?P<retention_days>\d+)",
+    }
+
+    SEP = ",\s*"
+
+    RE = re.compile(f"\({SEP.join([PARTSEGMENT_RE[s] for s in part_format])}\)")
+
+    match = RE.match(part_str)
```

Comment:
You are recompiling the regex every time you decode a part this way. I would actually move this outside out the the function and generate it for each storage for the sequence of PartSegment of each storage as they are constant.